File: server/src/collectors/sslblJp.js
/**
 * abuse.ch SSL Blacklist (SSLBL) — malicious SSL certificates.
 *
 * Public CSV is auth-free; we pull and surface entries whose listing host /
 * IP geolocates to JP-AS via the included AS info (when available).
 *
 * Endpoint:
 *   https://sslbl.abuse.ch/blacklist/sslipblacklist.csv    (IP:port + dst country)
 */

import { createThreatIntelCollector } from '../utils/threatIntelCollectorFactory.js';
import { TOKYO } from './_satelliteSeeds.js';

const URL_IP_CSV = 'https://sslbl.abuse.ch/blacklist/sslipblacklist.csv';
const FEODO_JSON = 'https://feodotracker.abuse.ch/downloads/ipblocklist.json';
const TIMEOUT_MS = 12000;

function parseCsv(text) {
  const out = [];
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith('#')) continue;
    const cols = line.split(',').map((c) => c.replace(/^"|"$/g, ''));
    if (cols.length < 4) continue;
    out.push({
      first_seen: cols[0],
      ip: cols[1],
      port: cols[2],
      malware: cols[3],
    });
  }
  return out;
}

// SSLBL csv has no country; intersect against Feodo's broader JSON list
// to flag JP-located rows.
async function fetchJpIps() {
  try {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), TIMEOUT_MS);
    const r = await fetch(FEODO_JSON, { signal: ctrl.signal });
    clearTimeout(t);
    if (!r.ok) return new Set();
    const arr = await r.json();
    const set = new Set();
    if (Array.isArray(arr)) {
      for (const row of arr) {
        if (String(row?.country || '').toUpperCase() === 'JP') set.add(String(row.ip_address));
      }
    }
    return set;
  } catch { return new Set(); }
}

export default createThreatIntelCollector({
  sourceId: 'sslbl',
  description: 'abuse.ch SSLBL — malicious-cert IPs intersected with JP-host set',
  run: async () => {
    let csv = '';
    let live = false;
    try {
      const ctrl = new AbortController();
      const t = setTimeout(() => ctrl.abort(), TIMEOUT_MS);
      const res = await fetch(URL_IP_CSV, { signal: ctrl.signal });
      clearTimeout(t);
      if (res.ok) {
        csv = await res.text();
        live = csv.length > 0;
      }
    } catch { /* fall through to seed mode */ }

    const rows = parseCsv(csv);
    const jpIps = await fetchJpIps();
    const jp = jpIps.size > 0 ? rows.filter((r) => jpIps.has(r.ip)) : [];
    const features = jp.map((r, i) => ({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: TOKYO },
      properties: {
        idx: i,
        first_seen: r.first_seen,
        ip: r.ip,
        port: r.port,
        malware: r.malware,
        source: live ? 'sslbl' : 'sslbl_seed',
      },
    }));
    return {
      features,
      source: live ? 'sslbl' : 'sslbl_seed',
      extraMeta: { total_rows: rows.length, jp_intersect: jpIps.size },
    };
  },
});

File: native/collectors/sources/sslbl_jp.c
/* collectors/cyber/sources/sslbl_jp.c
 * Port of server/src/collectors/sslblJp.js (createThreatIntelCollector).
 * Keyless. SSLBL IP CSV intersected with Feodo JP IP set, TOKYO points.
 * JS parseCsv is hand-rolled (split lines, split ',', strip ^"|"$, >=4 cols). */
#include "../../source.h"
#include "../../lib/threatintel.h"
#include "../../lib/feedlib.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define TOKYO_LON 139.6917
#define TOKYO_LAT 35.6895
#define URL_IP_CSV "https://sslbl.abuse.ch/blacklist/sslipblacklist.csv"
#define FEODO_JSON "https://feodotracker.abuse.ch/downloads/ipblocklist.json"

/* strip a single leading and trailing '"' in place (JS replace /^"|"$/g). */
static char *strip_q(char *s) {
  size_t n = strlen(s);
  if (n && s[n - 1] == '"') s[--n] = 0;
  if (s[0] == '"') return s + 1;
  return s;
}

/* trim ASCII whitespace in place, return start ptr. */
static char *trim(char *s) {
  while (*s == ' ' || *s == '\t' || *s == '\r' || *s == '\n') s++;
  size_t n = strlen(s);
  while (n && (s[n-1]==' '||s[n-1]=='\t'||s[n-1]=='\r'||s[n-1]=='\n')) s[--n]=0;
  return s;
}

static cJSON *run_fetch(const char *key, const source_ctx *ctx, void *ud) {
  (void)key; (void)ud;

  /* 1. Build the JP IP set from Feodo JSON (country === 'JP'). */
  cJSON *feodo = feed_get_json(ctx->http, FEODO_JSON, 12000);
  cJSON *jp_set = cJSON_CreateObject();   /* used as a string set */
  int jp_count = 0;
  if (cJSON_IsArray(feodo)) {
    cJSON *row;
    cJSON_ArrayForEach(row, feodo) {
      cJSON *c = cJSON_GetObjectItem(row, "country");
      const char *cc = (c && cJSON_IsString(c)) ? c->valuestring : "";
      if ((cc[0]=='J'||cc[0]=='j') && (cc[1]=='P'||cc[1]=='p') && !cc[2]) {
        cJSON *ipv = cJSON_GetObjectItem(row, "ip_address");
        if (ipv && cJSON_IsString(ipv) && ipv->valuestring[0]) {
          if (!cJSON_GetObjectItem(jp_set, ipv->valuestring)) {
            cJSON_AddBoolToObject(jp_set, ipv->valuestring, 1);
            jp_count++;
          }
        }
      }
    }
  }
  if (feodo) cJSON_Delete(feodo);

  /* 2. Fetch SSLBL CSV text. */
  char *csv = feed_get_text(ctx->http, URL_IP_CSV, 12000);
  cJSON *features = cJSON_CreateArray();
  if (csv && jp_count > 0) {
    int idx = 0;
    char *save = NULL;
    for (char *line = strtok_r(csv, "\n", &save); line;
         line = strtok_r(NULL, "\n", &save)) {
      char *t = trim(line);
      if (!*t || t[0] == '#') continue;
      /* split on ',', strip quotes, need >=4 cols. */
      char *cols[8]; int nc = 0;
      char *cs = NULL;
      for (char *tok = strtok_r(t, ",", &cs); tok && nc < 8;
           tok = strtok_r(NULL, ",", &cs)) {
        cols[nc++] = strip_q(tok);
      }
      if (nc < 4) continue;
      const char *first_seen = cols[0], *ip = cols[1],
                 *port = cols[2], *malware = cols[3];
      if (!cJSON_GetObjectItem(jp_set, ip)) continue;  /* jpIps.has(r.ip) */

      cJSON *f = cJSON_CreateObject();
      cJSON_AddStringToObject(f, "type", "Feature");
      cJSON *g = cJSON_CreateObject();
      cJSON_AddStringToObject(g, "type", "Point");
      cJSON *co = cJSON_CreateArray();
      cJSON_AddItemToArray(co, cJSON_CreateNumber(TOKYO_LON));
      cJSON_AddItemToArray(co, cJSON_CreateNumber(TOKYO_LAT));
      cJSON_AddItemToObject(g, "coordinates", co);
      cJSON_AddItemToObject(f, "geometry", g);

      cJSON *p = cJSON_CreateObject();
      cJSON_AddNumberToObject(p, "idx", idx);
      cJSON_AddStringToObject(p, "first_seen", first_seen);
      cJSON_AddStringToObject(p, "ip", ip);
      cJSON_AddStringToObject(p, "port", port);
      cJSON_AddStringToObject(p, "malware", malware);
      cJSON_AddStringToObject(p, "source", "sslbl");  /* live path only */
      cJSON_AddItemToObject(f, "properties", p);
      cJSON_AddItemToArray(features, f);
      idx++;
    }
  }
  if (csv) free(csv);
  cJSON_Delete(jp_set);
  return features;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  int n = threatintel_collect(ctx, sink, NULL, NULL, run_fetch, NULL);
  return n >= 0 ? 0 : -1;
}

static const source_def sslbl_jp_def = {
  .id = "sslbl-jp", .collector = "cyber",
  .name = "abuse.ch SSLBL (JP)", .name_ja = "abuse.ch SSLBL 日本",
   .update_interval_sec = 3600, .run = run };
REGISTER_SOURCE(sslbl_jp_def)
